iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -16,45 +17,163 @@
+                    it = m_instanceMap.erase(it);
+                    removedEntry = true;
+                    continue;
+                }
+                else if (removedEntry && it->second > removedValue)
+                {
+                    // update index due to removed element
+                    it->second--;
+                }
+                it++;
+            }
+            m_serviceDescriptionVector.erase(iterator);
+            removedElement = true;
+            // There can be not more than one element
+            break;
+        }
+        index++;
+        iterator++;
+    }
+
+    return removedElement;
 }
 
-void ServiceRegistry::find(InstanceSet_t& instances,
+void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
                            const CaproIdString_t& service,
-                           const CaproIdString_t& instance) const
+                           const CaproIdString_t& instance) const noexcept
 {
-    if (instance == iox::cxx::string<100>(capro::AnyInstanceString))
+    cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> intersection;
+
+    // Find (K1, K2)
+    // O(log n + log n + max(#PossibleServices + #possiblesInstances) + #intersection)
+    if (instance != capro::IdString_t(capro::AnyInstanceString)
+        && service != capro::IdString_t(capro::AnyServiceString))
     {
-        for (auto& instance : m_serviceMap[service].instanceSet)
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleServices;
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleInstances;
+
+        auto rangeServiceMap = m_serviceMap.equal_range(service);
+        for (auto entry = rangeServiceMap.first; entry != rangeServiceMap.second; ++entry)
+        {
+            possibleServices.push_back(entry->second);
+        }
+
+        auto rangeInstanceMap = m_instanceMap.equal_range(instance);
+        for (auto entry = rangeInstanceMap.first; entry != rangeInstanceMap.second; ++entry)
+        {
+            possibleInstances.push_back(entry->second);
+        }
+
+        ::std::set_intersection(possibleServices.begin(),
+                                possibleServices.end(),
+                                possibleInstances.begin(),
+                                possibleInstances.end(),
+                                ::std::back_inserter(intersection));
+
+        for (auto& value : intersection)
         {
-            instances.push_back(instance);
+            searchResult.push_back(m_serviceDescriptionVector[value]);
         }
     }
     else
     {
-        auto& instanceSet = m_serviceMap[service].instanceSet;
-        auto iter = std::find(instanceSet.begin(), instanceSet.end(), instance);
-        if (iter != instanceSet.end())
+        // Find (*, K2)
+        // O(log n + #result)
+        if (service == capro::IdString_t(capro::AnyServiceString)
+            && instance != capro::IdString_t(capro::AnyInstanceString))
+        {
```

Comment:
Please make this an `else_if`. It will reduce indentation and gives you 4 more character before clang-format wraps the line ;)

Reply:
Done